# References
- Blin, K., Shaw, S., Augustijn, H.E., Reitz, Z.L., Biermann, F., Alanjary, M., Fetter, A., Terlouw, B.R., Metcalf, W.W., Helfrich, E.J.N., van Wezel, G.P., Medema, M.H., Weber, T., 2023. antiSMASH 7.0: New and improved predictions for detection, regulation, chemical structures and visualisation. Nucleic Acids Research gkad344. [https://doi.org/10.1093/nar/gkad344](https://doi.org/10.1093/nar/gkad344)
- Bradnam, K.R., Fass, J.N., Alexandrov, A., Baranay, P., Bechner, M., Birol, I., Boisvert, S., Chapman, J.A., Chapuis, G., Chikhi, R., et al, 2013. Assemblathon 2: Evaluating de novo methods of genome assembly in three vertebrate species. GigaScience 2, 10. [https://doi.org/10.5524/100061](https://doi.org/10.5524/100061)
- Castillo-Alfonso, F., Valadez-Cano, C., Cejas-Añón, G., Utrilla, J., Sigala Alanis, J.-C., Le Borgne, S., Sales-Cruz, A.M., Vigueras-Ramírez, G., Olivares-Hernández, R., 2024. Sequencing, assembly, and genomic annotation of _Leucoagaricus gongylophorus_ LEU18496, a dikarya mutualistic species. Mol. Omics 20, 524–531. [https://doi.org/10.1039/D4MO00108G](https://doi.org/10.1039/D4MO00108G)
- Challis, R., Richards, E., Rajan, J., Cochrane, G., Blaxter, M., 2020. BlobToolKit – Interactive quality assessment of genome assemblies. G3 Genes|Genomes|Genetics 10, 1361–1374. [https://doi.org/10.1534/g3.119.400908](https://doi.org/10.1534/g3.119.400908)
- Earl, D.A., Bradnam, K.R., St John, J., Darling, A., Lin, D., Fass, J., Yu, H.O.K., Buffalo, V., Zerbino, D.R., Diekhans, M., et al, 2011. Assemblathon 1: A competitive assessment of de novo short read assembly methods. Genome Research 21, 2224–2241. [https://doi.org/10.1101/gr.126599.111](https://doi.org/10.1101/gr.126599.111)
- Hu, J., Fan, J., Sun, Z., Liu, S., 2020. NextPolish: A fast and efficient genome polishing tool for long-read assembly. Bioinformatics 36, 2253–2255. [https://doi.org/10.1093/bioinformatics/btz891](https://doi.org/10.1093/bioinformatics/btz891)
- Jackman, S.D., Vandervalk, B.P., Mohamadi, H., Chu, J., Yeo, S., Hammond, S.A., Jahesh, G., Khan, H., Coombe, L., Warren, R.L., Birol, I., 2017. ABySS 2.0: Resource-efficient assembly of large genomes using a Bloom filter. Genome Research 27, 768–777. [https://doi.org/10.1101/gr.214346.116](https://doi.org/10.1101/gr.214346.116)
- Käll, L., Krogh, A., Sonnhammer, E.L.L., 2004. A combined transmembrane topology and signal peptide prediction method. Journal of Molecular Biology 338, 1027–1036. https://doi.org/10.1016/j.jmb.2004.03.016
- Koren, S., Walenz, B.P., Berlin, K., Miller, J.R., Bergman, N.H., Phillippy, A.M., 2017. Canu: Scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation. Genome Research 27, 722–736. [https://doi.org/10.1101/gr.215087.116](https://doi.org/10.1101/gr.215087.116)
- Lanfear, R., Schalamun, M., Kainer, D., Wang, W., Schwessinger, B., 2019. MinIONQC: fast and simple quality control for MinION sequencing data. Bioinformatics 35, 523–525. [https://doi.org/10.1093/bioinformatics/bty654](https://doi.org/10.1093/bioinformatics/bty654)
- Liu, H., Wu, S., Li, A., Ruan, J., 2021. SMARTdenovo: a de novo assembler using long noisy reads. Gigabyte 2021, 1–9. [https://doi.org/10.46471/gigabyte.15](https://doi.org/10.46471/gigabyte.15)
- Luan, T., Commichaux, S., Hoffmann, M., Jayeola, V., Jang, J.H., Pop, M., Rand, H., Luo, Y., 2024. Benchmarking short and long read polishing tools for nanopore assemblies: achieving near-perfect genomes for outbreak isolates. BMC Genomics 25, 679. [https://doi.org/10.1186/s12864-024-10582-x](https://doi.org/10.1186/s12864-024-10582-x)
- Manni, M., Berkeley, M.R., Seppey, M., Simão, F.A., Zdobnov, E.M., 2021. BUSCO update: Novel and streamlined workflows along with broader and deeper phylogenetic coverage for scoring of eukaryotic, prokaryotic, and Viral genomes. Molecular Biology and Evolution 38, 4647–4654. [https://doi.org/10.1093/molbev/msab199](https://doi.org/10.1093/molbev/msab199)
- Merényi, Z., Krizsán, K., Sahu, N., Liu, X.-B., Bálint, B., Stajich, J.E., Spatafora, J.W., Nagy, L.G., 2023. Genomes of fungi and relatives reveal delayed loss of ancestral gene families and evolution of key fungal traits. Nat Ecol Evol 7, 1221–1231. [https://doi.org/10.1038/s41559-023-02095-9](https://doi.org/10.1038/s41559-023-02095-9)
- Naranjo‐Ortiz, M.A., Gabaldón, T., 2019. Fungal evolution: major ecological adaptations and evolutionary transitions. Biol Rev 94, 1443–1476. [https://doi.org/10.1111/brv.12510](https://doi.org/10.1111/brv.12510)
- Naranjo‐Ortiz, M.A., Gabaldón, T., 2019. Fungal evolution: diversity, taxonomy and phylogeny of the Fungi. Biological Reviews 94, 2101–2137. [https://doi.org/10.1111/brv.12550](https://doi.org/10.1111/brv.12550)
- Pryszcz, L.P., Gabaldón, T., 2016. Redundans: An assembly pipeline for highly heterozygous genomes. Nucleic Acids Research 44, e113. [https://doi.org/10.1093/nar/gkw294](https://doi.org/10.1093/nar/gkw294)
- Vaser, R., Sović, I., Nagarajan, N., Šikić, M., 2017. Fast and accurate de novo genome assembly from long uncorrected reads. Genome Res. 27, 737–746. [https://doi.org/10.1101/gr.214270.116](https://doi.org/10.1101/gr.214270.116)
- Wang, J., Chen, K., Ren, Q., Zhang, Y., Liu, J., Wang, G., Liu, A., Li, Y., Liu, G., Luo, J., Miao, W., Xiong, J., Yin, H., Guan, G., 2021. Systematic comparison of the performances of _de novo_ genome assemblers for Oxford Nanopore Technology reads from Piroplasm. Front. Cell. Infect. Microbiol. 11, 696669. [https://doi.org/10.3389/fcimb.2021.696669](https://doi.org/10.3389/fcimb.2021.696669)
- Zimin, A.V., Marçais, G., Puiu, D., Roberts, M., Salzberg, S.L., Yorke, J.A., 2013. The MaSuRCA genome assembler. Bioinformatics 29, 2669–2677. [https://doi.org/10.1093/bioinformatics/btt476](https://doi.org/10.1093/bioinformatics/btt476)